---
title: ActiveSupport::ErrorReporter
layout: default
---
<div class="main">
    <div class="banner">
        
            <span>Ruby on Rails 7.0.8</span><br />
        
        <div class="type">Class</div>
        <h1>
            ActiveSupport::ErrorReporter
            
                <span class="parent">&lt;
                    
                    <a href="../Object.html">Object</a>
                    
                </span>
            
        </h1>
        <ul class="files">
            
            <li><a href="../../files/activesupport/lib/active_support/error_reporter_rb.html">activesupport/lib/active_support/error_reporter.rb</a></li>
            
        </ul>
    </div>
    <div id="bodyContent">
        <div id="content">
  
    <div class="description">
      
<p><code>ActiveSupport::ErrorReporter</code> is a common interface for error reporting services.</p>

<p>To rescue and report any unhandled error, you can use the <code>handle</code> method:</p>

<pre><code>Rails.error.handle do
  do_something!
end
</code></pre>

<p>If an error is raised, it will be reported and swallowed.</p>

<p>Alternatively if you want to report the error but not swallow it, you can use <code>record</code></p>

<pre><code>Rails.error.record do
  do_something!
end
</code></pre>

<p>Both methods can be restricted to only handle a specific exception class</p>

<pre><code>maybe_tags = Rails.error.handle(Redis::BaseError) { redis.get(&quot;tags&quot;) }
</code></pre>

<p>You can also pass some extra context information that may be used by the error subscribers:</p>

<pre><code>Rails.error.handle(context: { section: &quot;admin&quot; }) do
  # ...
end
</code></pre>

<p>Additionally a <code>severity</code> can be passed along to communicate how important the error report is. <code>severity</code> can be one of <code>:error</code>, <code>:warning</code>, or <code>:info</code>. Handled errors default to the <code>:warning</code> severity, and unhandled ones to <code>:error</code>.</p>

<p>Both <code>handle</code> and <code>record</code> pass through the return value from the block. In the case of <code>handle</code> rescuing an error, a fallback can be provided. The fallback must be a callable whose result will be returned when the block raises and is handled:</p>

<pre><code>user = Rails.error.handle(fallback: -&gt; { User.anonymous }) do
  User.find_by(params)
end
</code></pre>

    </div>
  

  

  
  


  

  
    <h2 id="methods">Methods</h2>
    <ul>
      
        <li>
          <a href="#method-i-handle">handle</a>
        </li>
      
        <li>
          <a href="#method-c-new">new</a>
        </li>
      
        <li>
          <a href="#method-i-record">record</a>
        </li>
      
        <li>
          <a href="#method-i-report">report</a>
        </li>
      
        <li>
          <a href="#method-i-set_context">set_context</a>
        </li>
      
        <li>
          <a href="#method-i-subscribe">subscribe</a>
        </li>
      
    </ul>
  

  

  
    

    

    
      <!-- Section constants -->
      <h2 id="constants">Constants</h2>
      <table border='0' cellpadding='5'>
        
          <tr valign='top'>
            <td class="attr-name">SEVERITIES</td>
            <td>=</td>
            <td class="attr-value">%i(error warning info)</td>
          </tr>
          
        
      </table>
    

    
      <!-- Section attributes -->
      <h2 id="attributes">Attributes</h2>
      <table border='0' cellpadding='5'>
        
          <tr valign='top'>
            <td class='attr-rw'>
              [RW]
            </td>
            <td class='attr-name'>logger</td>
            <td class='attr-desc'></td>
          </tr>
        
      </table>
    

    <!-- Methods -->
    
      <h2 id="class-public-methods">Class Public methods</h2>
      
        <div class="method">
          <h3 id="method-c-new">
            
              new(*subscribers, logger: nil)
            
          </h3>

          
            <div class="description">
              
            </div>
          

          

          

          
          
            <details class="method__source">
              <summary>
                <span class="label">üìù Source code</span>
              </summary>

              <pre><code class="ruby"># File activesupport/lib/active_support/error_reporter.rb, line 46
    def initialize(*subscribers, logger: nil)
      @subscribers = subscribers.flatten
      @logger = logger
    end</code></pre>
              
                <a href="https://github.com/rails/rails/blob/fc734f28e65ef8829a1a939ee6702c1f349a1d5a/activesupport/lib/active_support/error_reporter.rb#L46" target="_blank" class="github_url">üîé See on GitHub</a>
              
            </details>
          
        </div>
        
      
    
      <h2 id="instance-public-methods">Instance Public methods</h2>
      
        <div class="method">
          <h3 id="method-i-handle">
            
              handle(error_class = StandardError, severity: :warning, context: {}, fallback: nil)
            
          </h3>

          
            <div class="description">
              <p>Report any unhandled exception, and swallow it.</p>

<pre><code>Rails.error.handle do
  1 + &#39;1&#39;
end
</code></pre>
            </div>
          

          

          

          
          
            <details class="method__source">
              <summary>
                <span class="label">üìù Source code</span>
              </summary>

              <pre><code class="ruby"># File activesupport/lib/active_support/error_reporter.rb, line 57
    def handle(error_class = StandardError, severity: :warning, context: {}, fallback: nil)
      yield
    rescue error_class =&gt; error
      report(error, handled: true, severity: severity, context: context)
      fallback.call if fallback
    end</code></pre>
              
                <a href="https://github.com/rails/rails/blob/fc734f28e65ef8829a1a939ee6702c1f349a1d5a/activesupport/lib/active_support/error_reporter.rb#L57" target="_blank" class="github_url">üîé See on GitHub</a>
              
            </details>
          
        </div>
        
        <div class="method">
          <h3 id="method-i-record">
            
              record(error_class = StandardError, severity: :error, context: {})
            
          </h3>

          
            <div class="description">
              
            </div>
          

          

          

          
          
            <details class="method__source">
              <summary>
                <span class="label">üìù Source code</span>
              </summary>

              <pre><code class="ruby"># File activesupport/lib/active_support/error_reporter.rb, line 64
    def record(error_class = StandardError, severity: :error, context: {})
      yield
    rescue error_class =&gt; error
      report(error, handled: false, severity: severity, context: context)
      raise
    end</code></pre>
              
                <a href="https://github.com/rails/rails/blob/fc734f28e65ef8829a1a939ee6702c1f349a1d5a/activesupport/lib/active_support/error_reporter.rb#L64" target="_blank" class="github_url">üîé See on GitHub</a>
              
            </details>
          
        </div>
        
        <div class="method">
          <h3 id="method-i-report">
            
              report(error, handled:, severity: handled ? :warning : :error, context: {})
            
          </h3>

          
            <div class="description">
              <p>When the block based <code>handle</code> and <code>record</code> methods are not suitable, you can directly use <code>report</code></p>

<pre><code>Rails.error.report(error, handled: true)
</code></pre>
            </div>
          

          

          

          
          
            <details class="method__source">
              <summary>
                <span class="label">üìù Source code</span>
              </summary>

              <pre><code class="ruby"># File activesupport/lib/active_support/error_reporter.rb, line 95
    def report(error, handled:, severity: handled ? :warning : :error, context: {})
      unless SEVERITIES.include?(severity)
        raise ArgumentError, &quot;severity must be one of #{SEVERITIES.map(&amp;:inspect).join(&quot;, &quot;)}, got: #{severity.inspect}&quot;
      end

      full_context = ActiveSupport::ExecutionContext.to_h.merge(context)
      @subscribers.each do |subscriber|
        subscriber.report(error, handled: handled, severity: severity, context: full_context)
      rescue =&gt; subscriber_error
        if logger
          logger.fatal(
            &quot;Error subscriber raised an error: #{subscriber_error.message} (#{subscriber_error.class})\n&quot; +
            subscriber_error.backtrace.join(&quot;\n&quot;)
          )
        else
          raise
        end
      end

      nil
    end</code></pre>
              
                <a href="https://github.com/rails/rails/blob/fc734f28e65ef8829a1a939ee6702c1f349a1d5a/activesupport/lib/active_support/error_reporter.rb#L95" target="_blank" class="github_url">üîé See on GitHub</a>
              
            </details>
          
        </div>
        
        <div class="method">
          <h3 id="method-i-set_context">
            
              set_context(...)
            
          </h3>

          
            <div class="description">
              <p>Update the execution context that is accessible to error subscribers</p>

<pre><code>Rails.error.set_context(section: &quot;checkout&quot;, user_id: @user.id)
</code></pre>

<p>See <code>ActiveSupport::ExecutionContext.set</code></p>
            </div>
          

          

          

          
          
            <details class="method__source">
              <summary>
                <span class="label">üìù Source code</span>
              </summary>

              <pre><code class="ruby"># File activesupport/lib/active_support/error_reporter.rb, line 88
    def set_context(...)
      ActiveSupport::ExecutionContext.set(...)
    end</code></pre>
              
                <a href="https://github.com/rails/rails/blob/fc734f28e65ef8829a1a939ee6702c1f349a1d5a/activesupport/lib/active_support/error_reporter.rb#L88" target="_blank" class="github_url">üîé See on GitHub</a>
              
            </details>
          
        </div>
        
        <div class="method">
          <h3 id="method-i-subscribe">
            
              subscribe(subscriber)
            
          </h3>

          
            <div class="description">
              <p>Register a new error subscriber. The subscriber must respond to</p>

<pre><code>report(Exception, handled: Boolean, context: Hash)
</code></pre>

<p>The <code>report</code> method <code>should</code> never raise an error.</p>
            </div>
          

          

          

          
          
            <details class="method__source">
              <summary>
                <span class="label">üìù Source code</span>
              </summary>

              <pre><code class="ruby"># File activesupport/lib/active_support/error_reporter.rb, line 76
    def subscribe(subscriber)
      unless subscriber.respond_to?(:report)
        raise ArgumentError, &quot;Error subscribers must respond to #report&quot;
      end
      @subscribers &lt;&lt; subscriber
    end</code></pre>
              
                <a href="https://github.com/rails/rails/blob/fc734f28e65ef8829a1a939ee6702c1f349a1d5a/activesupport/lib/active_support/error_reporter.rb#L76" target="_blank" class="github_url">üîé See on GitHub</a>
              
            </details>
          
        </div>
        
      
    
  
</div>

    </div>
</div>
